master/pkg/tasks/task_spec.go

```diff
@@ -1,57 +1,266 @@
 package tasks
 
 import (
+	"archive/tar"
 	"crypto/tls"
 	"encoding/json"
+	"fmt"
+	"path/filepath"
+	"strconv"
+	"strings"
 
+	"github.com/docker/docker/api/types/mount"
+
+	"github.com/determined-ai/determined/master/pkg/container"
+	"github.com/determined-ai/determined/master/pkg/etc"
 	"github.com/determined-ai/determined/master/pkg/workload"
 
 	"github.com/determined-ai/determined/master/pkg/archive"
 	"github.com/determined-ai/determined/master/pkg/device"
 	"github.com/determined-ai/determined/master/pkg/model"
 )
 
-// TaskSpec provides the necessary information for an agent to start a task.
+// InnerSpec defines the interface for a particular kind of task container. TODO: Think of a better
+// name.
+type InnerSpec interface {
+	// Archives returns the files to include in the container for this task (apart from the base files
+	// put into in all containers).
+	Archives(*model.AgentUserGroup) []container.RunArchive
+	// Description returns a brief description of this task.
+	Description() string
+	// Entrypoint returns the command and arguments to run in the container for this task.
+	Entrypoint() []string
+	// Environment returns the container environment for this task.
+	Environment(TaskSpec) model.Environment
+	// EnvVars returns the environment variables to set for this task (apart from the base ones set for
+	// all containers).
+	EnvVars(TaskSpec) map[string]string
+	// LoggingFields returns fields to include in each record of structured (i.e., Fluent Bit) logging.
+	LoggingFields() map[string]string
+	// Mounts returns the list of Docker mounts to use for this task.
+	Mounts() []mount.Mount
+	// ShmSize specifies the shared memory size to allocate to this task's container in bytes (0 for
+	// default behavior).
+	ShmSize() int64
+	// UseFluentLogging specifies whether to use Fluent Bit logging (as opposed to native logging).
+	UseFluentLogging() bool
+	// UseHostMode indicates whether host mode networking would be desirable for this task.
+	UseHostMode() bool
+}
+
+// This alias allows TaskSpec to privately embed the public InnerSpec so that it can reuse (some of)
+// its methods without directly providing access to the field to other packages.
+type inner = InnerSpec
```

Comment:
blocking: it seems the golang type alias are not meant for everyday use: [reference](https://talks.golang.org/2016/refactor.article#TOC_5.3.). It should be fine to use `InnerSpec` directly.

Reply:
Hmm, I don't get the impression that that's what to take from that page. The original motivation for type aliases may have come from package refactoring, but no reputable sources I can find are saying that they should _only_ be used then (I only found two random [blog](https://alehatsman.com/posts/golang_type_declaration_abuse.html) [posts](https://yourbasic.org/golang/type-alias/), both also citing the same page). In this case, a type alias has _exactly_ the properties we want: it allows us to not have to retype a bunch of methods while not exporting the embedded member. If there is a technical reason not to do this, I'd be happy to hear it.

Reply:
There is no technical reason not to do this. It's just that type alias were invented to solve the graduate code repair case. See [Go1.9](https://golang.org/doc/go1.9#language) and [Proposal: Type Aliases
](https://go.googlesource.com/proposal/+/master/design/18130-type-alias.md). I think Go is designed to be transparent and it uses only the upper cases to distinguish the methods you want to export from other methods. But it seems not to fulfill all the needs we have in this case. Since I couldn't find anywhere in the Go documentation that explicitly says the use cases beyond gradual refactoring are not recommended, I'll approve this PR.